dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java

```diff
@@ -664,6 +664,10 @@ public void removeAllBundles(Context context, Item item) throws AuthorizeExcepti
         while(bundles.hasNext())
         {
             Bundle bundle = bundles.next();
+            //Mark bitstreams to be deleted from all bundles that are about to be deleted
+            for(Bitstream bs : bundle.getBitstreams()){
+                bs.setDeleted(true);
+            }
```

Comment:
I'm not sure this is the right "level" to be flagging Bitstreams as deleted. Having this at the item Level seems odd, and it shouldn't *only* exist in `removeAllBundles()` (as Bitstreams can be deleted individually as well).

In 5.x, Bitstreams were set as deleted in the `BitstreamStorageManager.delete()` method: https://github.com/DSpace/DSpace/blob/dspace-5_x/dspace-api/src/main/java/org/dspace/storage/bitstore/BitstreamStorageManager.java#L547

Deleted bitstreams were then cleaned up / expunged in `BitstreamStorageManager.cleanup()`: https://github.com/DSpace/DSpace/blob/dspace-5_x/dspace-api/src/main/java/org/dspace/storage/bitstore/BitstreamStorageManager.java#L570

In 6.x, this BitstreamStorageManager was refactored into a new [BitstreamStorageService:](https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/storage/bitstore/BitstreamStorageServiceImpl.java)  and a [DSBitstoreService](https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/storage/bitstore/DSBitStoreService.java).  It looks like the `cleanup()` now occurs in `BitstreamStorageService`, but the `delete()` is in `DSBitstoreService`.  However, from the looks of it, the `delete()` now removes the file instead of flagging as deleted -- I'm not sure where the 6.x is flagging deletions now (or if it even is).

Reply:
Very enlightening thank you. 

To clarify, I'm under the assumption that the expected behavior specifically when expunging an item (not explicitly deleting a bitstream specifically) is that all it's respective bundles & bitstreams are deleted and / or in the bitstream's case marked for deletion. Then by running `cleanup()` BitstreamStorageService would generate a `List<Bitstream>` 's to be removed based on the `deleted` column in the Bitstream table, then removed the respective bitstreams ( https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/storage/bitstore/BitstreamStorageServiceImpl.java#L229 ). Specifically from the context of expunging an item I am not seeing that item's respective bitstreams are being marked for deletion, therefore when `cleanup()` is ran those respective bitstreams are pass over and simply left orphaned. I know I have personally tested this behavior against DSpace Community dsapce-6_x (latest) alongside all my 6.x clients and so has Chris Wilper. I've also just recreated the behavior again as a sanity check :) .

When a bitstream is explicitly deleted from the Admin Edit -> Item Bitstreams page I can confirm when looked up in the Bitstream table that it is marked for deletion. But Bitstreams are not marked for deletion via expunging the items.

As for the code comment:
I believe the the flagging of the bitstreams to be deleted is in a convenient spot as `removeAllBundles` is only ever invoked through the expunge item process and we're already iterating through all the item's respective bundles at this point so the code is very low impact.
But if you prefer I can modify the `cleanup()` process to include all Bitstreams that do not have their UUID's present in all possible table locations (bundle2bitstream, community, etc).

Reply:
If bitstream deletion works properly from the Admin Edit -> Item Bitstreams, I'd recommend debugging to see how that page is correctly marking bitstreams for deletion.  Ideally, marking bitstreams for deletion should happen in *one place in the code*, instead of doing it multiple ways in different methods. (i.e. Duplicating the same logic to different methods / layers of the API isn't recommended.) 

That's why I'm not sure the `removeAllBundles()` options is the best solution -- it sounds to me like this "works" in some scenarios, but not in all. So, that implies to me that the issue is not in `removeAllBundles()` itself, but rather the code that marks bitstreams as deleted is not being triggered in all deletion scenarios.